stl/inc/semaphore

```diff
@@ -0,0 +1,304 @@
+
+        // release thread: Increment semaphore counter, then load waiting counter;
+        // acquire thread: Increment waiting counter, then load semaphore counter;
+
+        // memory_order_seq_cst for all four operations guarantees that either release thread loads
+        // the incremented value, or acquire thread loads the incremented value, or both, but not none.
+        // memory_order_seq_cst might be superfluous for some hardware mappings of C++ memory model,
+        // but from the point of C++ memory model itself it is needed, weaker orders don't work.
+
+        const ptrdiff_t _Prev = _Counter.fetch_add(static_cast<ptrdiff_t>(_Update), memory_order_seq_cst);
+        _STL_VERIFY(_Prev + _Update > 0 && _Prev + _Update <= _Least_max_value,
+            "Precondition: update <= max() - counter (N4861 [thread.sema.cnt]/8)");
+
+        const ptrdiff_t _Waiting_upper_bound = _Waiting.load(memory_order_seq_cst);
+
+        if (_Waiting_upper_bound == 0) {
+            // Definitely no one is waiting
+        } else if (_Waiting_upper_bound <= _Update) {
+            // No more waiting threads than update, can wake everyone.
+            _Counter.notify_all();
+        } else {
+            // Wake at most _Update. Though repeated notify_one() is somewhat less efficient than single notify_all(),
+            // the amount of OS calls is still the same; the benefit from trying not to wake unnecessary threads
+            // is expected to be greater than the loss on extra calls and atomic operations.
+            for (ptrdiff_t _Thd = _Update; _Thd != 0; --_Thd) {
+                _Counter.notify_one();
+            }
+        }
+    }
+
+    void _Wait(const unsigned long _Remaining_timeout) noexcept {
+        // See the comment in release()
+        _Waiting.fetch_add(1, memory_order_seq_cst);
+        ptrdiff_t _Current = _Counter.load(memory_order_seq_cst);
+        if (_Current == 0) {
+            __std_atomic_wait_direct(&_Counter, &_Current, sizeof(_Current), _Remaining_timeout);
+        }
+        _Waiting.fetch_sub(1, memory_order_relaxed);
+    }
+
+    void acquire() noexcept /* strengthened */ {
+        ptrdiff_t _Current = _Counter.load(memory_order_relaxed);
+        for (;;) {
+            while (_Current == 0) {
+                _Wait(_Atomic_wait_no_timeout);
+                _Current = _Counter.load(memory_order_relaxed);
+            }
+            _STL_VERIFY(_Current > 0 && _Current <= _Least_max_value,
+                "Invariant: counter >= 0, and counter <= max() "
+                "possibly caused by preconditions violation (N4861 [thread.sema.cnt]/8)");
+
+            if (_Counter.compare_exchange_weak(_Current, _Current - 1, memory_order_seq_cst)) {
+                break;
+            }
+        }
+    }
+
+    _NODISCARD bool try_acquire() noexcept {
+        ptrdiff_t _Current = _Counter.load(memory_order_relaxed);

```

Comment:
This one certainly can't be relaxed since upon seeing 0 we need to have syncronized with another thread that has made it be 0. We're seq_cst everywhere else, we probably should be here.

Reply:
Ditto for the other acquires.

Reply:
No, the CAS below synchronizes with, the purpose of this is to have best guess for the first CAS attempt

Reply:
There isn't a cas attempt here if _Current was 0. For example:

```
semaphore s; //assume some thread will call release somewhere
while (!s.try_acquire()) ;
```

This can become an infinite loop because relaxed means the thing couold be enregistered etc.

I think anywhere we say relaxed needs a fairly large explaination of where synchronization is being provided and a proof that that is safe. Whereas dropping to effectively acq_rel is far less suspect.